### PR TITLE
AC: Added support for BiDAF model

### PR DESCRIPTION
**Pull request by Anna-Mironova**

### CODE REVIEW

File: tools/accuracy_checker/accuracy_checker/annotation_converters/README.md [hunk 1]
```diff
@@ -268,6 +268,8 @@ The main difference between this converter and `super_resolution` in data organi
   * `max_query_length` - maximum number of tokens for the question (Optional, default value is 64).
   * `doc_stride` -stride size between chunks for splitting up long document (Optional, default value is 128).
   * `lower_case` - allows switching tokens to lower case register. It is useful for working with uncased models (Optional, default value is False)
+* `squad_bidaf` - converts the Stanford Question Answering Dataset ([SQuAD](https://rajpurkar.github.io/SQuAD-explorer/)) to `QuestionAnsweringBiDAFAnnotation`. **Note:** This converter not only converts data to metric specific format but also tokenize and encodes input for BiDAF.
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/annotation_converters/README.md:271]:
```suggestion
* `squad_bidaf` - converts the Stanford Question Answering Dataset ([SQuAD](https://rajpurkar.github.io/SQuAD-explorer/)) to `QuestionAnsweringBiDAFAnnotation`. **Note:** This converter not only converts data to metric specific format but also tokenize and encodes input for BiDAF using nltk.word_tokenize
```

File: tools/accuracy_checker/accuracy_checker/annotation_converters/__init__.py [hunk 2]
```diff
@@ -152,5 +153,5 @@
     'CriteoKaggleDACConverter',
     'FeaturesRegressionConverter',
     'Im2latexDatasetConverter',
-
+    'SQUADConverterBiDAF',
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/annotation_converters/__init__.py:156]: please move to other squad converters

File: tools/accuracy_checker/accuracy_checker/annotation_converters/squad_bidaf.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+        annotations = []
+        for example in examples:
+            cw, cc = self._preprocess_text(example.context_text, True)
+            qw, qc = self._preprocess_text(example.question_text)
+            cw_idx_in_context = self._get_tokens_indexes_in_context(example.context_text, cw.reshape(-1))
+            identifier = ['context_word_{}'.format(0),
+                          'context_char_{}'.format(0),
+                          'query_word_{}'.format(0),
+                          'query_char_{}'.format(0)]
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/annotation_converters/squad_bidaf.py:144]: why it is always 0?
should be idx of example in enumerate(examples)

File: tools/accuracy_checker/accuracy_checker/metrics/question_answering.py [hunk 4]
```diff
@@ -97,8 +98,8 @@ def reset(self):
 class ExactMatchScore(PerImageEvaluationMetric):
     __provider__ = 'exact_match'
 
-    annotation_types = (QuestionAnsweringAnnotation,)
-    prediction_types = (QuestionAnsweringPrediction,)
+    annotation_types = (QuestionAnsweringAnnotation, QuestionAnsweringBiDAFAnnotation, )
+    prediction_types = (QuestionAnsweringPrediction, QuestionAnsweringBiDAFPrediction, )
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/metrics/question_answering.py:102]: do we really need new data type for prediction?

File: tools/accuracy_checker/accuracy_checker/adapters/nlp.py [hunk 5]
```diff
@@ -196,7 +196,34 @@ def process(self, raw, identifiers=None, frame_meta=None):
...
 
+class QuestionAnsweringBiDAFAdapter(Adapter):
+    __provider__ = 'bidaf_question_answering'
+    prediction_types = (QuestionAnsweringPrediction, )
+
+    @classmethod
+    def parameters(cls):
+        parameters = super().parameters()
+        parameters.update({
+            'start_pos': StringField(description="Output layer name for answer start position."),
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/adapters/nlp.py:207]: please add out suffix for emphasis that you expect output layer name, not an index

File: tools/accuracy_checker/accuracy_checker/annotation_converters/squad_bidaf.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                        for answer in answers:
+                            answer.update({
+                                "text": preprocess(answer["text"])
+                            })
+
+                    example = SquadExample(
+                        qas_id=qas_id,
+                        question_text=question_text,
+                        context_text=context_text,
+                        answer_text=answer_text,
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/annotation_converters/squad_bidaf.py:86]: this varaiable is always None in your code

File: tools/accuracy_checker/accuracy_checker/annotation_converters/squad_bidaf.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            indexes.append(idx + offset)
+            offset += idx + len(w)
+            rem = rem[idx + len(w):]
+        return indexes
+
+    def convert(self, check_content=False, progress_callback=None, progress_interval=100, **kwargs):
+        examples = self._load_examples(self.testing_file)
+
+        annotations = []
+        for (example_index, example) in enumerate(examples):
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/annotation_converters/squad_bidaf.py:137]:
```suggestion
        for example_index, example in enumerate(examples):
```

File: tools/accuracy_checker/accuracy_checker/launcher/onnx_launcher.py [hunk 8]
```diff
@@ -141,6 +141,8 @@ def fit_to_input(self, data, layer_name, layout, precision):
             return data.astype(input_precision)
         if len(np.shape(data)) == 5 and len(layout) == 5:
             return np.transpose(data, layout).astype(input_precision)
+        if (len(np.shape(data)) == 3 and len(layer_shape) == 2) or (len(np.shape(data)) == 5 and len(layer_shape) == 4):
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/launcher/onnx_launcher.py:144]:
```suggestion
        if len(np.shape(data)) -1 == len(layer_shape):
```

File: tools/accuracy_checker/accuracy_checker/postprocessor/extract_answers_tokens.py [hunk 9]
```diff
@@ -181,3 +184,38 @@ def _strip_spaces(text):
 
         output_text = orig_text[orig_start_position:(orig_end_position + 1)]
         return output_text
+
+class ExtractSQUADPredictionBiDAF(Postprocessor):
+
+    __provider__ = 'bidaf_extract_answers_tokens'
+
+    annotation_types = (QuestionAnsweringBiDAFAnnotation, )
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/postprocessor/extract_answers_tokens.py:192]: can this postprocessor is also work with QuestionAnsweringAnnotation?

File: tools/accuracy_checker/accuracy_checker/annotation_converters/squad.py [hunk 10]
```diff
@@ -99,7 +99,7 @@ def __init__(
         self.char_to_word_offset = char_to_word_offset
 
         # Start and end positions only has a value during evaluation.
-        if start_position_character is not None and not is_impossible:
+        if start_position_character is not None and not is_impossible and answer_text is not None:
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/annotation_converters/squad.py:102]: in which situation answer_text is not None?

Comment (eaidova) [reply tools/accuracy_checker/accuracy_checker/annotation_converters/squad.py:102]: please check that your changes does not break bert models evaluation

Comment (Anna-Mironova) [reply tools/accuracy_checker/accuracy_checker/annotation_converters/squad.py:102]: bert models work well with these changes

**Review by eaidova — APPROVED** [review]

**Pull request merged**